### PR TITLE
Improve message on parse error

### PR DESCRIPTION
**Pull request by duelafn**
Hello, forgive the unsolicited, drive-by, necromantic pull-request (all the evils), but I found your repo and was quite pleased to find a way to incorporate the use of `terminals_expected` with a higher-level interface of Marpa (I was doing something much uglier).

In my opinion, one of the better uses of `terminals_expected` is improving the quality of parse error messages, which is what this pull-request adds.

BEFORE:

    $ perl calc.pl '2+3'
    No lexeme found at 0 at calc.pl line 92.

AFTER:

    $ perl calc.pl '2+3'
    Parse error on line 1 at '2+3', expecting: print varname

NOTE: Any literal in the grammar (e.g., "('print')") will produce less than stellar error messages since they are given internal token names. That is why I moved "('print')" to its own token. I left the rest for illustration and to avoid needlessly expanding the code.

EXAMPLE OF BAD INLINE LITERALS (first three are probably internal names for '(', '"', and '='):

    $ perl calc.pl 'print'
    Parse error on line 1 at ';', expecting: [Lex-0] [Lex-10] [Lex-3] number varname

I see that this repo is for a presentation that you gave/give and error messages may not be the point of that talk, so feel free to drop this PR if you like. I was just so happy to find an example which fits in with my error processing (examples I've seen all show just `$grammar->parse()`) that I felt I should send a comment/patch showing just how easy it is with Marpa.

(LEGAL: Feel free to use this code however you like, patch submitted into the public domain or CC0)

### CODE REVIEW

**Comment by choroba** [comment]
Thanks for reminding me about the repository. I'm adding some more links, one of them will be [Karel::Parser](https://metacpan.org/source/Karel::Parser) which is more recent than this repo and uses exactly `terminals_expected` for error reporting.

File: calc.pl [hunk 1]
```diff
@@ -15,7 +15,7 @@
 Statement  ::= Assign separ                action => none
              | Output separ                action => none
 Assign     ::= Var ('=') Expression        action => store
-Output     ::= ('print') List              action => show
+Output     ::= print List                  action => show
```
Comment (choroba) [inline calc.pl]: Keep the parentheses, just remove the quotes

File: calc.pl [hunk 2]
```diff
@@ -62,7 +64,7 @@
 sub none        {}
 sub single      { $_[1] }
 sub numify      { 0 + $_[1] }
-sub show        { say $_[1] }
+sub show        { say $_[2] }
```
Comment (choroba) [inline calc.pl]: ... so you don't have to change this.

File: calc.pl [hunk 3]
```diff
@@ -89,8 +91,25 @@
...
+        PARSE_ERROR($input, $recce->pos, $recce->terminals_expected);
     }
 }
 $recce->value;
 
+
+sub PARSE_ERROR {
+    my ($input, $pos, $expected) = @_;
+
+    my $line = 1+(substr($input,0,$pos) =~ tr/\n/\n/);
```
Comment (choroba) [inline calc.pl]: `tr/\n//` is enough.

**Comment by duelafn** [comment]
Oh, looked at Karel::Parser and yeah, that's nice. I tweaked this example for shortness and clarity (and newly discovered line_column) in case you want to include it, but as i said, no heartbreak if you don't. (not sure proper convention for modifying a PR, whether I should rebase into a single clean patch or leave it separate, I would normally rebase, but it's gone public so I avoid rebasing after it goes public)

Thanks for the links!

**Comment by choroba** [comment]
Improved and merged as c15abd8f2e9347398bc6a0eb4402654d39d9b7f4
Thanks for the PR.

**Pull request closed without merging**